grouper/graph.py

```diff
@@ -436,6 +444,62 @@ def _get_edges(session):
 
         return edges
 
+    def _get_grants_by_permission(
+        self,
+        graph,  # type: DiGraph
+        group_grants,  # type: Dict[str, List[GroupPermissionGrant]]
+        service_account_grants,  # type: Dict[str, List[ServiceAccountPermissionGrant]]
+    ):
+        # type: (...) -> AllGrants
+        """Build a map of permissions to users and service accounts with grants."""
+        service_grants = defaultdict(
+            lambda: defaultdict(set)
+        )  # type: Dict[str, Dict[str, Set[str]]]
+        for account, service_grant_list in iteritems(service_account_grants):
+            for service_grant in service_grant_list:
+                service_grants[service_grant.permission][account].add(service_grant.argument)
+
+        # For each group that has a permission grant, determine all of its users from the graph,
+        # and then record each permission grant of that group as a grant to all of those users.
+        # This ensures that each permission grant is recorded for all affected users exactly once.
```

Comment:
Teeny tiny nit: the "exactly" comes from the use of sets rather than the algorithm (as the comment sort of implies) I believe

Reply:
Thanks, fixed!